src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -622,45 +623,45 @@ discover_daos_mount(void)
 	num_dfs++;
 }
 
+#define MNT_TYPE_FUSE	"fuse.daos"
 static void
 discover_dfuse(void)
 {
 	FILE          *fp;
-	const char     mnt_type_fuse[] = "fuse.daos";
 	struct mntent *fs_entry;
-	char          *pt_end;
+	struct dfs_mt *pt_dfs_mt;
 
 	num_dfs = 0;
 
 	fp = setmntent("/proc/self/mounts", "r");
-	/* daos_init() is not called yet, so we can not use D_WARN() etc. */
 
 	if (fp == NULL) {
-		D_ERROR("failed to open file: /proc/self/mounts\n");
+		D_ERROR("failed to open /proc/self/mounts: %d (%s)\n", errno, strerror(errno));
 		return;
 	}
 
 	while ((fs_entry = getmntent(fp)) != NULL) {
-		if (strncmp(fs_entry->mnt_type, mnt_type_fuse, sizeof(mnt_type_fuse)) == 0) {
-			dfs_list[num_dfs].dfs_dir_hash = NULL;
-			dfs_list[num_dfs].len_fs_root  = strnlen(fs_entry->mnt_dir, DFS_MAX_PATH);
-			if (dfs_list[num_dfs].len_fs_root >= DFS_MAX_PATH) {
-				D_WARN("mnt_dir[] is too long! Skip this entry.");
+		pt_dfs_mt = &dfs_list[num_dfs];
```

Comment:
This function will of course fail if there are more than 8 dfuse instances on a node which is far from impossible.  It'll also fail in a way that overwrites memory which isn't desirable, a simple check on num_dfs after incrementing it would be the quickest fix.

That also makes me wonder what would happen for dfuse instances that are running where this code has no access to the mount point, one option would be to call stat() on them here and skip them if the stat() fails, the other option would be to add them to the array as happens now but then it'll presumably fail on first access.

Not a regression but this is a trivially reproducible memory corruption issue - that could potentially be triggered by other users on the node.

Reply:
The range of num_dfs is checked now. access() is called to check whether read permission is granted before adding the dfuse mount point into dfs_list. Thank you!